repository: use "best match" for supplychain/delivery selection

cirocosta opened the pull request:
## Changes proposed by this PR

1. implement a function (`BestLabelMatches`) that, given something with labels and a set of objects that have selectors, gives the subset of the latter that "best matches" that initial label set according to [RFC-0015].

2. make use of such function for both delivery and workload

[RFC-0015]: https://github.com/vmware-tanzu/cartographer/pull/319

closes #298 

## Release Note

Allow users to have their Workload and Deliverables selecting ClusterSupplyChain and ClusterDelivery (accordingly) based on multiple labels rather than exact match.


## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [ ] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- [ ] Filled in the [Release Note](#Release-Note) section above
- [ ] Modified the docs to match changes <!-- TBD: reference doc editing guidance -->

emmjohnson commented on pkg/controller/deliverable/reconciler_test.go:
```diff
@@ -268,7 +268,7 @@ var _ = Describe("Reconciler", func() {
 					},
 				}
-				repo.GetDeliveriesForDeliverableReturns([]v1alpha1.ClusterDelivery{delivery}, nil)
+				repo.GetDeliveriesForDeliverableReturns([]*v1alpha1.ClusterDelivery{&delivery}, nil)
```
indentation

cirocosta replied:
sorry, wdym? is the suggestion to split the args into multiple lines? the change here was solely making `clustedelivery` a pointer

emmjohnson replied:
it was showing all the way to left before? maybe bad diff?

emmjohnson commented on pkg/controller/workload/reconciler_test.go:
```diff
@@ -267,7 +267,7 @@ var _ = Describe("Reconciler", func() {
 					},
 				}
-				repo.GetSupplyChainsForWorkloadReturns([]v1alpha1.ClusterSupplyChain{supplyChain}, nil)
+				repo.GetSupplyChainsForWorkloadReturns([]*v1alpha1.ClusterSupplyChain{&supplyChain}, nil)
```
indentation

cirocosta commented:
force push: squashed & rebase

cirocosta commented:
working on fixing the conflicts

cirocosta commented:
force-push: rebase + fix conflicts

emmjohnson commented on pkg/repository/repository.go:
```diff
@@ -214,23 +214,27 @@ func (r *repository) patchUnstructured(ctx context.Context, existingObj *unstruc
-		if selectorMatchesLabels(supplyChain.Spec.Selector, workload.Labels) {
-			clusterSupplyChains = append(clusterSupplyChains, supplyChain)
-		}
+	selectorGatters := []SelectorGetter{}
```
selectorGetters instead of Gatters?

Commit:
controller: use best match sc/delivery selection

repository: add BestLabelMatches function

	based on [RFC-0015], this function implements a "best label
	metching" algorithm that tries to find the object that has a
	Selector that fits the best a bigen object's label set according
	to the followign rules:

		1. labels are fully contained in the selector
		2. the most number of label matches, the better
		3. the most specific the selector, the better

	note that despite those rules, we can still return more than one
	match - it's then up to a higher level function to decide what
	to do with that (thus, preserving the current behavior).

controller/supchain: use of best label matching

	rather than sticking to exact matches, leverage the new
	"bestmatch" algo for supplychains.

controller/delivery: use of best label matching

	rather than sticking to exact matches, leverage the new
	"bestmatch" algo for deliveries.

[RFC-0015]: https://github.com/vmware-tanzu/cartographer/pull/319

Signed-off-by: Ciro S. Costa <ciroscosta@vmware.com>
